crates/rethnet_evm/src/blockchain/remote.rs

```diff
@@ -0,0 +1,152 @@
+            return Ok(Some(block));
+        }
+
+        if let Some(block) = self
+            .runtime
+            .block_on(self.client.get_block_by_hash_with_transaction_data(hash))?
+        {
+            self.fetch_and_cache_block(cache, block).map(Option::Some)
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Retrieves the block with the provided number, if it exists.
+    pub fn block_by_number(&self, number: &U256) -> Result<Arc<DetailedBlock>, RpcClientError> {
+        let cache = self.cache.upgradable_read();
+
+        if let Some(block) = cache.block_by_number(number).cloned() {
+            Ok(block)
+        } else {
+            let block = self.runtime.block_on(
+                self.client
+                    .get_block_by_number_with_transaction_data(BlockSpec::Number(*number)),
+            )?;
+
+            self.fetch_and_cache_block(cache, block)
+        }
+    }
+
+    /// Retrieves the block that contains a transaction with the provided hash, if it exists.
+    pub fn block_by_transaction_hash(
+        &self,
+        transaction_hash: &B256,
+    ) -> Result<Option<Arc<DetailedBlock>>, RpcClientError> {
+        if let Some(block) = self
+            .cache
+            .read()
+            .block_by_transaction_hash(transaction_hash)
+            .cloned()
+        {
+            Ok(Some(block))
+        } else if let Some(transaction) = self
+            .runtime
+            .block_on(self.client.get_transaction_by_hash(transaction_hash))?
+        {
+            // TODO: is this true?
+            self.block_by_hash(&transaction.block_hash.expect("Not a pending transaction"))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Retrieves the total difficulty at the block with the provided hash.
+    pub fn total_difficulty_by_hash(&self, hash: &B256) -> Result<Option<U256>, RpcClientError> {
+        let cache = self.cache.upgradable_read();
+
+        if let Some(difficulty) = cache.total_difficulty_by_hash(hash).cloned() {
+            Ok(Some(difficulty))
+        } else if let Some(block) = self
+            .runtime
+            .block_on(self.client.get_block_by_hash_with_transaction_data(hash))?
+        {
+            let total_difficulty = block
+                .total_difficulty
+                .expect("Must be present as this is not a pending transaction");
+
+            self.fetch_and_cache_block(cache, block)?;
+
+            Ok(Some(total_difficulty))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Fetches detailed block information and caches the block.
+    fn fetch_and_cache_block(
+        &self,
+        cache: RwLockUpgradableReadGuard<'_, SparseBlockchainStorage>,
+        block: remote::eth::Block<remote::eth::Transaction>,
+    ) -> Result<Arc<DetailedBlock>, RpcClientError> {
+        let total_difficulty = block
+            .total_difficulty
+            .expect("Must be present as this is not a pending transaction");
```

Comment:
Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/1b0a93b7ecea339e84905e861b4b6068c7951438

A pending block does not contain several fields, including `total_difficulty`. Since this is guaranteed to not be a pending block, we can call `expect` on the `Option`.